lib/line-index/src/lib.rs

```diff
@@ -115,29 +125,34 @@ impl LineIndex {
         LineIndex { newlines, line_wide_chars }
     }
 
+    /// Transforms the `TextSize` into a `LineCol`.
```

Comment:
Let's add `assert` that `offset` is inbounds and document that this might panic for invalid offsets (potentially, we might even want to panic for offsets pointing int he miidle of multibyte character).

Reply:
what about trying to avoid panics altogether, but instead doing various checks for validity (like the one you mentioned) and returning option/result if failed? then callers can choose what to do

Reply:
:thinking: so this is indexing API, and indexing API idiomatically panics. We might also do things like assert _in debug_ that column does not split a codepoint, and just skip this validation in release. The contract here is in general the same as for `&str` indexing --- you can only use indexes which you derived from this same string. 

The twist though, is that, unlike strings, the caller doesn't have an easy way to validate the index, should they choose to do so. And, given that line-columns often come from external sources, you do have to validate. 

I think I convinced myself that the current API is actaully the right one. For `TextSize`, which is the native encoding, we can have the contract that the index is trusted, and, if it actually comes from an external source, the caller can check manually using `is_codepoint_boundary` or how the corresponding `&str` method is called. 

For line col, we need to return an option. 

That is, line/col and text size are not symmetric, one is native and the other is not.

Reply:
if we want to assert that offset is in-bounds, i think we also need to track the overall len of the text that this LineIndex is for. i don't see how i can recover that currently.

i added a check that the offset isn't in the middle of a multibyte char

Reply:
> i think we also need to track the overall len of the text that this LineIndex is for. i don't see how i can recover that currently.

I think the most elegant way is to change `newlines.push(TextSize::from(0));` to pushing `text.len()` as the last emelent instead of the first element and adjusting `.partition_point` condition accordinagly (`test_every_chars` slow test should prevent off-by-one errors)

Reply:
i'm not sure that'll work with the current setup. since we'll incorrectly say that when offset=len, the linecol is a non-existent line at column 0. see current failed tests.

Reply:
i changed it to just have the size be another field, which lets us avoid an unwrap anyway